walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,88 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.SortedSet;
+import java.util.TreeSet;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() > 1440)
+      return Collections.emptyList();
+
+    SortedSet<TimeRange> mandatoryBlockers = new TreeSet<>(TimeRange.ORDER_BY_START);
+
+    for (Event event : events) {
+      for (String attendee : event.getAttendees()) {
+        if (request.getAttendees().contains(attendee)) {
+          // this is a blocker, split up the available time segment
+          mandatoryBlockers.add(event.getWhen());
+        }
+      }
+    }
+
+    List<TimeRange> combinedMandatoryBlockers = mergeTimeRanges(mandatoryBlockers);
+    return subtractTimeRanges(combinedMandatoryBlockers, request.getDuration());
+  }
+
+  private static List<TimeRange> mergeTimeRanges(Collection<TimeRange> ranges) {
+    List<TimeRange> merged = new ArrayList<>(ranges.size());
+    TimeRange prev = null;
+
+    for (TimeRange current : ranges) {
+      if (prev == null) {
+        prev = current;
+        merged.add(prev);
+        continue;
+      }
+
+      if (prev.end() > current.start()) {
+        if (current.end() > prev.end()) {
+          prev = TimeRange.fromStartEnd(prev.start(), current.end(), false);
+          merged.set(merged.size() - 1, prev);
+        }
+      } else {
+        prev = current;
+        merged.add(prev);
+      }
+    }
+
+    return merged;
+  }
+
+  /**
+   * Subtracts the time ranges given by `ranges` from the time ranges representing the whole day.
+   * `ranges` must be sorted by start time and not contain any overlapping ranges.
+   */
+  private static List<TimeRange> subtractTimeRanges(List<TimeRange> ranges,
```

Comment:
This function name is a bit misleading. Consider naming it findAvailability(conflictingTimes, minDuration)